### PR TITLE
wolves_added

### PR DESCRIPTION
**Pull request by Top1Miami**
Wolveeeeeeees

### CODE REVIEW

File: ITMO_FS/wrappers/randomized/TPhMGWO.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        Parameters
+        ----------
+        wolfNumber : integer
+        	Number of search agents used to find solution for features selection problem
+		seed : integer
+			Random seed used to initialize np.random.seed()
+		alpha : float
+			weight of importance of classification accuracy
+			Note alpha is used in equation that counts fitness as fitness = alpha * score + beta * |selected_features| / |features| where alpha = 1 - beta
+		knn : KNeighborsClassifier
```
[the remark this answers is not in the document]
Comment (Top1Miami) [reply ITMO_FS/wrappers/randomized/TPhMGWO.py]: But the actual paper says it should

[the remark this answers is not in the document]
Comment (LastShekel) [reply ITMO_FS/wrappers/randomized/TPhMGWO.py]: it doesn't have to be exactly knn
just set knn as default model, but change model to parameter

File: ITMO_FS/utils/information_theory.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from math import exp
+from math import log
+
+
+def calc_conditional_entropy(x_j, y):
+    Kx = max(x_j)
+    countX = np.zeros(Kx + 1)
+    dictYByX = {}
+    for i in range(Kx):
+        dictYByX.update({i + 1: {}})
```
Comment (LastShekel) [inline ITMO_FS/utils/information_theory.py:11]: dictYByX={i+1:{} for i in range(Kx)}

File: ITMO_FS/utils/information_theory.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+def elog(x):
+    if x <= 0. or x >= 1.:
+        return 0
+    else:
+        return x*log(x)
+
+def calc_entropy(x):
+	d = dict()
+	for obj in x:
+		d[obj] = d.get(obj, 0) + 1
```
Comment (LastShekel) [inline ITMO_FS/utils/information_theory.py:52]: numpy hist

File: ITMO_FS/utils/information_theory.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        countX[x_val] += 1
+        dictYByX[x_val].update({y_val : dictYByX[x_val].get(y_val, 0) + 1})
+    entropy = 0.0
+    for i in range(Kx + 1):
+        if countX[i] == 0:
+            continue
+        partEntropy = 0.0
+        curDict = dictYByX[i]
+        for inClass in curDict.values():
+            partEntropy += elog(inClass / countX[i])
```
Comment (LastShekel) [inline ITMO_FS/utils/information_theory.py:24]: redo with numpy
actualy you could call function over numpy array

File: ITMO_FS/filters/multivariate/DISRwithMassive.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+			calc_entropy(list(zip(x_i, y))) - calc_entropy(list(zip(x_j, y))) + calc_entropy(list(zip(x_i, x_j, y)))
+
+	def __chained_information(self, x_i, x_j, y):
+		return mutual_information(x_i, y) + mutual_information(x_j, y) + self.__complementarity(x_i, x_j, y)
+
+	def __count_weight(self, i):
+		temp_difference = 0
+		for j in range(self.n_features):
+			temp_difference += self.edges[i][j] * self.vertices[i] * self.vertices[j]
+			temp_difference += self.edges[j][i] * self.vertices[i] * self.vertices[j]
```
Comment (LastShekel) [inline ITMO_FS/filters/multivariate/DISRwithMassive.py:52]: you could multiply by 2
take a look at numpy multiply

File: ITMO_FS/filters/multivariate/DISRwithMassive.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		free_features = np.array([], dtype=np.integer)
+		self.selected_features = np.arange(self.n_features, dtype=np.integer)
+		self.vertices = np.ones(self.n_features)
+		self.edges = np.zeros((self.n_features, self.n_features))
+		for i in range(self.n_features):
+			for j in range(self.n_features):
+				entropy = calc_entropy(list(zip(X[:, i], X[:, j])))
+				if entropy == 0.:
+					self.edges[i][j] = 0.
+				else:
```
Comment (LastShekel) [inline ITMO_FS/filters/multivariate/DISRwithMassive.py:85]: just check if entropy != 0.:

File: ITMO_FS/filters/multivariate/MultivariateFilter.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+from .measures import GLOB_MEASURE
+from ...utils import generate_features
+
+import numpy as np
+
+class MultivariateFilter(object):
```
Comment (LastShekel) [inline ITMO_FS/filters/multivariate/MultivariateFilter.py:6]: this stuff have to be harf tested

File: ITMO_FS/utils/information_theory.py (hunk 3 above)
Comment (LastShekel) [reply ITMO_FS/utils/information_theory.py:52]: X could be a list of tuples, leave it

**Pull request merged**